### PR TITLE
Filter nodes 2 [WIP]

### PR DESCRIPTION
**Pull request by foxish**
Fixes https://github.com/apache-spark-on-k8s/spark/issues/73#issuecomment-284100998

This should be the last change before freezing.
Will test the change and update the PR.

@ash211 @mccheah

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/NodePortUrisDriverServiceManager.scala [hunk 1]
```diff
@@ -16,9 +16,11 @@
  */
 package org.apache.spark.deploy.rest.kubernetes
 
-import io.fabric8.kubernetes.api.model.{Service, ServiceBuilder}
-import scala.collection.JavaConverters._
+import com.fasterxml.jackson.annotation.{JsonIgnoreProperties, JsonProperty}
+import com.fasterxml.jackson.databind.ObjectMapper
+import io.fabric8.kubernetes.api.model.{Node, Service, ServiceBuilder}
 
+import scala.collection.JavaConverters._
```
Comment (iyanuobidele) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/NodePortUrisDriverServiceManager.scala:23]: fix headers

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/NodePortUrisDriverServiceManager.scala [hunk 2]
```diff
@@ -63,6 +67,27 @@ private[spark] class NodePortUrisDriverServiceManager extends DriverServiceManag
     require(nodeUrls.nonEmpty, "No nodes found to contact the driver!")
     nodeUrls
   }
+
+  private def nodeHasUnschedulableTaint(node: Node): Boolean = {
+    val schedulerTaints = node.getMetadata.getAnnotations
+      .get(SCHEDULER_TAINTS_ANNOTATION)
+    schedulerTaints match {
+      case null =>
+        return false
```
Comment (iyanuobidele) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/NodePortUrisDriverServiceManager.scala:76]: nit: remove return

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/NodePortUrisDriverServiceManager.scala [hunk 3]
```diff
@@ -63,6 +67,27 @@ private[spark] class NodePortUrisDriverServiceManager extends DriverServiceManag
...
+
+      case json =>
+        try {
+          val mapper = new ObjectMapper()
+          val taints = mapper.readValue(json, classOf[Array[SchedulerAnnotation]])
+          return taints.exists(t => t.key == SCHEDULER_MASTER_TAINT_KEY
```
Comment (iyanuobidele) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/NodePortUrisDriverServiceManager.scala:82]: ditto

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/NodePortUrisDriverServiceManager.scala [hunk 4]
```diff
@@ -63,6 +67,27 @@ private[spark] class NodePortUrisDriverServiceManager extends DriverServiceManag
...
+            && t.effect == SCHEDULER_MASTER_TAINT_EFFECT)
+        } catch {
+          case e: Throwable =>
+            logWarning(s"Failed to parse taints on node ${node.getMetadata.getName}", e)
+        }
+    }
+    false
```
Comment (iyanuobidele) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/NodePortUrisDriverServiceManager.scala:89]: you can move the false keyword after the logWarning

**Comment by iyanuobidele** [comment]
I'm sure the tests failed cos of the headers.

I see that If the json deserialization fails, you're defaulting to false ? Is this always the case ?

LGTM after the nits

**Comment by ash211** [comment]
Please rebase onto `branch-2.1-kubernetes` and send this PR into that branch instead of `k8s-support-alternate-incremental` which is now deprecated.

**Comment by ash211** [comment]
@foxish are you able to pick this back up again or should we close it?

If so, maybe with 1.6 released the taint was agreed upon and we can put it into the fabric8 client and then use here.

**Comment by foxish** [comment]
Sorry, this fell off my plate. There is more agreement now on the taint. I'll update this PR. https://github.com/apache-spark-on-k8s/spark/issues/73 is still tracking it.

**Comment by mccheah** [comment]
This is likely out of date with V2 submission work, @foxish ?

**Comment by foxish** [comment]
Yes, closing this. Will revisit if the issue is raised in future.

**Pull request closed without merging**